Add "importRuntime" option to "transform-runtime"

nicolo-ribaudo opened the pull request:
> When "importRuntime" is true, regenerator doesn't assume that
> `regeneratorRuntime` is available globally, but it injects an `import`
> statement (or `require`, depending on the source type)

Example output:
```js
var _regeneratorRuntime = require("regenerator-runtime");

var _marked =
/*#__PURE__*/
_regeneratorRuntime.mark(f);

function f() {
  return _regeneratorRuntime.wrap(function f$(_context) {
    while (1) switch (_context.prev = _context.next) {
      case 0:
      case "end":
       return _context.stop();
    }
  }, _marked, this);
}
```

I had to update the Babel deps because I hit https://github.com/babel/babel/issues/8508 and https://github.com/babel/babel/issues/8722.

nicolo-ribaudo committed:
Update babel deps

nicolo-ribaudo committed:
Add "importRuntime" option to "transform-runtime"

When "importRuntime" is true, regenerator doesn't assume that
`regeneratorRuntime` is available globally, but it injects an `import`
statement (or `require`, depending on the source type)

nicolo-ribaudo committed:
Update README

nicolo-ribaudo commented on package.json:
```diff
@@ -28,21 +28,21 @@
-    "regenerator-transform": "^0.13.3",
+    "regenerator-preset": "file:packages/regenerator-preset",
+    "regenerator-runtime": "file:packages/regenerator-runtime",
+    "regenerator-transform": "file:packages/regenerator-transform",
```
This changes were made by lerna I think.

benjamn replied:
Should we revert the lerna changes?

benjamn commented on packages/regenerator-transform/src/util.js:
```diff
@@ -23,10 +25,30 @@ export function getTypes() {
-export function runtimeProperty(name) {
+const runtimeNames = new WeakMap();
+
+export function runtimeProperty(name, scope, opts) {
```
Now that `runtimeProperty` takes options and does more than just construct a member expression, I think we should move it out of `util.js` and make it a method of the `Emitter` class. That way we won't need to pass the options explicitly, since they're stored as a property of the `Emitter` object now. From what I can tell, everywhere `runtimeProperty` was previously called, we have access to an `Emitter` instance, so this change should be pretty straightforward. What do you think @nicolo-ribaudo?

benjamn reviewed (changes requested):
This seems like the right direction, though I have one substantive suggestion.

nicolo-ribaudo committed:
Move to emitter and revert lerna changes

benjamn committed:
Add package-lock.json to regenerator-preset package.

benjamn committed:
Add package-lock.json to regenerator-transform package.

benjamn committed:
Update package-lock.json.

This removes the file:... versions from package-lock.json, which I believe
have been causing Travis CI tests to fail for PR #356.

loganfsmyth commented:
To throw it out there, I don't have a problem with landing this if you want, but I don't think it's something we'd recommend people use on the Babel side since `plugin-transform-runtime` also handles injecting a reference to the runtime and is already very recommended in Babel 7.

nicolo-ribaudo commented:
Closing since this is now handled by `@babel/plugin-transform-runtime`
